Update to beanqueen salad

beanqueen opened the pull request.

netme commented on CHANGES:
```diff
@@ -1,9 +1,22 @@
-0.4.14-beanqueen
+0.5
```
Maybe 0.6 as we have discussed before.

skoczen replied:
0.5 should be fine.

netme commented on salad/__init__.py:
```diff
@@ -1 +1 @@
-VERSION = "0.4.14-beanqueen-1"
+VERSION = "0.5"
```
The same versioning issue.

netme commented on salad/cli.py:
```diff
@@ -8,37 +9,34 @@
 
 BROWSER_CHOICES = [browser.lower()
-                   for browser in DesiredCapabilities.__dict__.keys()
+                   for browser in list(DesiredCapabilities.__dict__.keys())
```
Why do we need to convert explicitly to list?

beanqueen replied:
this was part of making it compatible for python 3, but i think we can do without the .keys() and it will do the same

netme commented on salad/cli.py:
```diff
@@ -49,17 +47,21 @@ def main(args=sys.argv[1:]):
+    world.remote_capabilities['version'] = parsed_args.browserversion
+    world.remote_capabilities['platform'] = parsed_args.platform
+    world.remote_capabilities['trustAllSSLCertificates'] = True
+    world.remote_capabilities['acceptSslCerts'] = True
```
I prefer to do it this way:

```python
world.remote_capabilities = {
    'version': parsed_args.browserversion,
    'platform': parsed_args.platform,
    'trustAllSSLCertificates': True,
    'acceptSslCerts': True
}
```

skoczen replied:
+1 to this format,

Also, I'd strongly prefer that adding security exceptions (like trusting all SSL certs) is an explicit choice by the user.  We shouldn't be disabling security by default.

beanqueen replied:
ok, i'll move it to a program option --trust-all-ssl

netme commented on salad/cli.py:
```diff
@@ -67,10 +69,26 @@ def main(args=sys.argv[1:]):
+        scenarios = set()
+        for part in parsed_args.scenarios.split(','):
+            x = part.split('-')
+            scenarios.update(list(range(int(x[0]), int(x[-1])+1)))
```
`list` is a bit slower than `[]`:

```
scenarios.update([range(int(x[0]), int(x[-1])+1)])
```

beanqueen replied:
but pylint --py3k will complain: "range-builtin-not-iterating" if we use []

netme commented on salad/features/steps-browser-browsers.feature:
```diff
@@ -11,6 +11,6 @@ Feature: Ensuring that javascript works in all supported browsers
     Examples:
     | browser   |
-    | chrome    |
+    #| chrome    | chrome does not work on travis at the moment
```
Then please remove it :)

skoczen replied:
I am -1 to salad's being built around travis.

beanqueen replied:
what do you suggest? is there a travis alternative? we also shouldn't build just because "it works on my computer" which it does with chrome

santagada replied:
I can try to if it out on travis if possible.

netme commented on salad/steps/browser/alerts.py:
```diff
@@ -1,59 +1,127 @@
+        alert = world.browser.driver.switch_to_alert()
+        # switch_to_alert() always returns an object, but if there is no
+        # alert, the following line will raise an exception
+        alert.text
```
What's the reason for this statement?

skoczen replied:
+1

beanqueen replied:
i wanted to explain why i had to add the line "alert.text", although the text is not further used.
get_alert() used to through the exception here, but driver.switch_to_alert() does not
and i also did not want to change the logic that is using the exception

netme commented on salad/steps/browser/alerts.py:
```diff
@@ -1,59 +1,127 @@
+        # switch_to_alert() always returns an object, but if there is no
+        # alert, the following line will raise an exception
+        alert.text
     except:
```
You should specify which exact exception you are catching.

netme commented on salad/steps/browser/alerts.py:
```diff
@@ -1,59 +1,127 @@
-@step(r'enter "([^"]*)" into the prompt')
+@step(r'enter "([^"]*)" into the prompt$')
 def enter_into_the_prompt(step, text):
     if not hasattr(world, "prompt") or not world.prompt:
```
```python
if not getattr(world, 'prompt', None):
```

netme commented on salad/steps/browser/alerts.py:
```diff
@@ -1,59 +1,127 @@
-    world.prompt.fill_with(text)
+@step(r'enter the stored value of "([^"]*)" into the prompt$')
+def enter_stored_value_into_prompt(step, name):
+    if not hasattr(world, "prompt") or not world.prompt:
```
```python
if not getattr(world, 'prompt', None):
```

netme commented on salad/steps/browser/alerts.py:
```diff
@@ -1,59 +1,127 @@
+@step(r'(?:store|remember) the( lowercase| uppercase)? text of the '
+      '(?:alert|prompt) as "([^"]+)"$')
+def store_text_of_element_with_case_option_as(step, upper_lower, name):
+    if not hasattr(world, "prompt") or not world.prompt:
```
```python
if not getattr(world, 'prompt', None):
```

netme commented on salad/steps/browser/browsers.py:
```diff
@@ -1,11 +1,51 @@
+
+    if with_timestamp:
+        ts = datetime.now().strftime('%Y-%m-%d_%H:%M:%S')
+        ts = '_%s' % (ts, )
```
Better:
```python
now = datetime.now()
timestamp = now.strftime('_%Y-%m-%d_%H:%M:%S')
```

netme commented on salad/steps/browser/finders.py:
```diff
@@ -22,49 +33,59 @@
-    return ele
-
+    # to support the splinter legacy functions
+    legacy_functions = ['by_value', 'by_partial_href', 'by_href']
```
If `legacy_functions` is not changeable, better to use a tuple:

```python
legacy_functions = ('by_value', 'by_partial_href', 'by_href')
```

netme commented on salad/steps/browser/javascript.py:
```diff
@@ -1,21 +1,40 @@
+    def assert_javascript_returns_value(negate, script, value):
+        try:
+            assert_equals_with_negate(
+                "%s" % world.browser.evaluate_script(script), value, negate)
```
` "%s" % world.browser.evaluate_script(script)` = `str(world.browser.evaluate_script(script))`

netme commented on salad/steps/browser/javascript.py:
```diff
@@ -1,21 +1,40 @@
-        logger.info("Attempted to run javascript in a javascript-disabled browser. Moving along.")
+        wait_for_completion(wait_time, assert_javascript_returns_value,
+                            negate, script, value)
+    except:
```
Please catch the proper exception.

netme commented on salad/tests/html/browser/js.html:
```diff
@@ -2,10 +2,17 @@
     <head>
         <title>Javascript title</title>
         <script>
```
type="text/javascript"

netme commented on salad/tests/util.py:
```diff
@@ -17,3 +29,81 @@ def assert_with_negate(assertion, negator=None):
+        if length <= 3:
+            return name
+        index = randint(1, len(name)-2)
+        return name[:index] + ' ' + name[index+1:]
```
```python
def generate_content(type_of_fill, length):
    random_string = generate_random_string(length)
    if type_of_fill == 'email':
        return '{}@example.org'.format(random_string)
    elif type_of_fill == 'string':
        return random_string
    elif type_of_fill == 'name':
        return ''.join([random_string] * 2)
```

netme commented:
Special thanks for PEP8 :)

skoczen commented:
Hi Jana and all,

For big releases like this, the best way to do it is to have the development happen out in the open, so all parties can review and contribute before things are “done.”

Having a giant PR come in proposed as the next major release with no context is a bit frightening!

That said, I also recognize this situation is special - I haven’t actively contributed to salad for quite some time, and you all are the ones really driving development.  Looking through the PR, I’m a bit concerned that it’s being locked into particular stacks (travis, browserstack), and that there are things in there that shouldn’t be (someone’s travis keys are in there, for example).

But more critically, I’m not honestly sure of what all the proposed changes are, and don’t see documentation in the README or elsewhere.  Could someone put together a clear list of the changes introduced in this PR, especially noting any backwards-incompatable changes?

I’d love to start there, get all of us clear on what's here and what's not, and then get this moved forward and shipped out when it’s ready!

Posting this comment on the PR, as well - let’s continue the discussion there!

-Steven

skoczen commented:
Hey Jana,

Yep, I saw that - but there doesn’t appear to be the corresponding changes to the salad documentation, so new users know how to use all the new features.

Also, at least at my first glance, these two are backwards incompatible:
- added $ at the end of regexes which from now on disallows wrong usage of steps
- Removed zope support

Saucelabs - (sorry, meant that instead of browserstack) - didn’t know that on the API keys, that’s great.

Also, let’s keep this conversation going here:
https://github.com/salad/salad/pull/60

Thanks!

skoczen commented:
Also, to be clear, I'm very much +1 to better code and more functionality - I just want to make sure we get it out there in a way that's consistent with how the project has been put out there, and provides a clear transition for existing users. :)

beanqueen commented:
what is this PR doing generally?
--------------------------------
- cleanup code
- add a lot of functionality that came up while using salad on a daily basis
- using selenium directly instead of via splinter because it offers so much more (splinter is still there, though, so all the nice short splinter functions can still be used)
- also read the CHANGES files for a more complete list of new stuff


two things are backwards incompatible:
-------------------------------------
- added $ at the end of regexes which from now on disallows wrong usage of steps:
We had some people misuse steps. so let's say they used a step that matched with 'And I should see the element with the id "foo"'.
But they thought they could add stuff to the end of a step and that this stuff would actually be tested as well, for example 'And I should see the element with the id "foo" and it should have the text "bar"'.
So the people wanted to test that foo has the text bar, but they only found out that foo was actually present on the page.
The regexes did not have a $ at the end, so it was possible to add arbitrary text at the end of them.
This is no longer allowed and we could quickly find all the wrongly used steps in our features.

- Removed zope support:
There was a problem with zope. No version of it could be installed any longer with the current salad requirements. 
It was impossible to install salad. removing zope fixed this problem. Does anybody use zope? Can someone get it back to work?


new in travis.yml:
-------------------
We are now running tests locally on travis and remotely on saucelabs. saucelabs has a free service for open source projects. I generated a key for this project (salad/salad). it is hashed somewhere in the travis.yml
Some features cannot be run with certain browsers or browser+OS combinations either due to bugs or because some functionality is not implemented for the browser.
these tests are skipped for the respective browser/browser+OS combination. also at the moment, travis has problems with chrome, so the tests are not run on travis with chrome.


should be split this big PR into several smaller ones?
------------------------------------------------------
i think the smallest sane RP is up to https://github.com/salad/salad/commit/32983e9bd83b64e32c43b4d101e46d998b0b6cdb
because the stuff has been code reviewed and it's been running for a while on our computers without problems.
BUT and this is the big BUT: the tests were not passing there. not because the PRs were bad, but because the travis.yml had to be changed completely.

It took quite a while to make this work. there are memory issues with 'lettuce', so travis tests also failed because the travis gave up.
it would be better to switch to lettuce 0.3.0, but it's not on master yet and there are still some bugs in it. we're using it in the company now from my fork.

so anyway, commits kept adding up while trying to get the tests to pass. this is basically all i tried after this commit. trying to make tests pass.
fixing bugs in salad, finding out about bugs in chrome/travis/lettuce/firefox/.., adding tests for untested steps, rewriting tests to be more clear, and of course adding saucelabs and phantomjs tests.

i would not split up this PR, but someone else can do it if necessary.

skoczen commented:
This is awesome, thanks for the big list! Makes it much easier to parse and understand!

Will take a good look through this tonight and/or tomorrow, and get comments in. Thanks for all the work on this, and the PR!

skoczen commented on .travis.yml:
```diff
@@ -1,36 +1,92 @@
   global:
-    - PATH=$PATH:/tmp/bin
+  - secure: NYFGMEkVVvBx8NLGcWnAkieJ11vxT4EVqN/9iQ95IF2XiswtkmUnFFRGyH1KbwaftzbXsg6S1gc79WxR2/EGBJuIHtFVsb8A7r6vedkX3dhcyjITa41uSP5EWANYdl6nsf4/u5oINXoGf/InI6jwqBpRhNV+ruq08937ZBZACek=
+  - secure: TU9KI3phk22RCMMCPrSYQ7W3KY56MKJXLs0+ajYXunvP93OKOMG6XUVkkoHjMH4cm9tSe4kH64PhET8GWxJbZw8U5+5F7GEiA490Dn4D2FpwxZI+W7v2M5g1hQeSQjXnWPVhmDvkDaUvkdt97k+qJe5wdsVFdiTB9vg3JLXLOyg=
```
Two of these are listed with the same key.  Imagine only one is being used?

skoczen commented on .travis.yml:
```diff
@@ -1,36 +1,92 @@
+  - BROWSER="phantomjs" FEATURE="salad/features/steps-browser-page.feature"
+  - BROWSER="phantomjs" FEATURE="salad/features/steps-browser-remember.feature"
 install:
-  - "python setup.py install"
```
So, when I read this, tests no longer run at all without saucelabs. Is that right?

skoczen commented on README.md:
```diff
@@ -1,5 +1,7 @@
 Salad is a helpful mix of some great BDD packages like lettuce and splinter, seasoned with some common modules.  Its goal is to make writing acceptance tests downright fun.
 
+Salad was originally developed at [Wieden+Kennedy](http://wk.com), and was subsequently [open-sourced](https://github.com/wieden-kennedy/salad) under the BSD license.
```
It's weird that this shows up - what branch/commit was this PR based on?

skoczen commented on salad/cli.py:
```diff
@@ -8,37 +9,34 @@
                         help=('Browser to use. Options: %s Default is %s.' %
                               (BROWSER_CHOICES, DEFAULT_BROWSER)))
 
+    parser.add_argument('--browserversion', help=('Browser version to use.'))
```
Choices and/or example?a

skoczen commented on salad/cli.py:
```diff
@@ -8,37 +9,34 @@
                                      description=("BDD browswer-automation "
                                                   "made tasty."))
 
+    parser.add_argument('-V', action='store_true', default=False,
```
would prefer this as `--version`, so it's not confused with the more common `-v` == verbose

skoczen commented on salad/cli.py:
```diff
@@ -49,17 +47,21 @@ def main(args=sys.argv[1:]):
     parser.add_argument('--timeout',
```
Just saucelabs?

skoczen commented on salad/cli.py:
```diff
@@ -8,37 +9,34 @@
+                        help=('Platform to use. Options: %s Default is %s.' %
+                              (PLATFORM_CHOICES, DEFAULT_PLATFORM)))
+
     parser.add_argument('--remote-url',
```
An example would be great.  Does this work with any selenum url?

medwards replied:
Yup! This one should date back to when I was still involved.

On 2 February 2015 at 11:31, Steven Skoczen <notifications@github.com>
wrote:

> In salad/cli.py
> <https://github.com/salad/salad/pull/60#discussion_r23916562>:
>
> >                          help=('Browser to use. Options: %s Default is %s.' %
> >                                (BROWSER_CHOICES, DEFAULT_BROWSER)))
> >
> > +    parser.add_argument('--browserversion', help=('Browser version to use.'))
> > +
> > +    parser.add_argument('--platform', default=DEFAULT_PLATFORM,
> > +                        help=('Platform to use. Options: %s Default is %s.' %
> > +                              (PLATFORM_CHOICES, DEFAULT_PLATFORM)))
> > +
> >      parser.add_argument('--remote-url',
>
> An example would be great. Does this work with any selenum url?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/salad/salad/pull/60/files#r23916562>.
>

skoczen commented on salad/cli.py:
```diff
@@ -67,10 +69,26 @@ def main(args=sys.argv[1:]):
+    if parsed_args.timeout:
         world.remote_capabilities['idle-timeout'] = parsed_args.timeout
+    else:
+        world.remote_capabilities['idle-timeout'] = 120
```
This is fine, but totally just curious: this code looks to be exactly equivalent to the old code. Why the change?

skoczen commented on salad/cli.py:
```diff
@@ -67,10 +69,26 @@ def main(args=sys.argv[1:]):
+        world.remote_capabilities['idle-timeout'] = 120
+
+    # travis job number
+    if os.environ.get('TRAVIS_JOB_NUMBER'):
```
This feels like a weird place to have custom, travis-specific code.

skoczen commented on salad/features/salad_terrains.py:
```diff
@@ -14,6 +14,9 @@
 from salad.logger import logger
 
 
+LETTUCE_TEMP_FILE = '/tmp/temp_lettuce_test'
```
Better: use `tempfile`: https://docs.python.org/2/library/tempfile.html

skoczen commented on salad/features/salad_terrains.py:
```diff
@@ -51,10 +54,13 @@ def teardown_test_server(total):
 
 
 @after.all
 def remove_tempfile(total):
```
If python tempfile is used, this shouldn't be necessary.

skoczen commented on salad/features/steps-browser-blur.feature:
```diff
@@ -0,0 +1,19 @@
+Feature: Make sure the focus and blur steps work
+
+
+    Scenario: 17. Focusing and blurring works
```
Unclear why the scenarios are all numbered now.   Seems like it's going to add maintenance overhead keeping those numbers straight.  Rationale?

skoczen commented on salad/features/steps-browser-remember.feature:
```diff
@@ -0,0 +1,216 @@
+          And I look around
+         # --- string with length
+         # --- test element contains string / element is string
+         When I store a random <what> of length 20 as "string_with_length"
```
Interesting new feature.  Not totally sure what I think about it, but really interesting to see salad developing in this way!

skoczen commented on salad/other_features/steps-browser-alerts.feature:
```diff
@@ -0,0 +1,152 @@
+Feature: Ensuring that the alert/prompt steps work
```
Why was this moved to "other_features"?

skoczen commented:
Just about done with CR, have to run - will get the rest done, and big-pictures thoughts in when I'm back!

skoczen commented on salad/tests/util.py:
```diff
@@ -17,3 +29,81 @@ def assert_with_negate(assertion, negator=None):
         assert assertion
+
+
+def assert_value(type_of_match, value, text, negate):
```
The name of this feels odd to me.  `assert_match`, perhaps?

skoczen commented on salad/tests/util.py:
```diff
@@ -17,3 +29,81 @@ def assert_with_negate(assertion, negator=None):
+        raise NotImplementedError(msg)
+
+
+def transform_for_upper_lower_comparison(stored, current, upper_lower):
```
This whole pattern should be simplified to something like

"contains foo" - no transformation
"contains foo, case-insensitive"  - both the match and the key are `.lower()`-cased, and compared.  Upper should not exist, and if the default is case-sensitive, no need to explicitly set it.

+1 to changing the storage function to the above, and simplifying this function.

skoczen commented on sauce_connect_setup.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
Would prefer things like this live in a `scripts` folder, instead of directly in root.

skoczen commented on setup.py:
```diff
@@ -6,20 +6,18 @@
 SOURCE_DIR = os.path.join(ROOT_DIR)
 
-requirements = ["nose==1.3.0", "splinter>=0.4.9", "lettuce>=0.2.19"]
+requirements = ["nose", "splinter", "lettuce"]
```
-1.  

Version-number specification is a very good thing, and prevents salad from randomly breaking if one of those packages introduces a backwards-incompatible release.

skoczen commented on setup.py:
```diff
@@ -6,20 +6,18 @@
-    author_email="qa@deliveryhero.com",
-    url="https://github.com/beanqueen/salad",
+    author="Jana Rekittke",
+    author_email="jana@rekittke.name ",
```
The author and author_email should actually stay (I must have missed this in previous PRs) as

```python
author="Steven Skoczen",
author_email="steven@inkandfeet.com",
```

Please note that this is authorship of the entire project, and who people should contact for support, not just this PR.

skoczen commented on setup.py:
```diff
@@ -6,20 +6,18 @@
     version=VERSION,
-#    download_url = ['https://github.com/gabrielfalcao/lettuce/tarball/fork', ],
     install_requires=requirements,
-#    dependency_links = ['https://github.com/gabrielfalcao/lettuce/tarball/fork#egg=lettuce-0.2.19', ],
```
The reason for a fork of lettuce being used was that the default lettuce was destroying people's *production* databases.  I haven't followed lettuce in a couple years (and have no idea if the maintainer has listened to the thousands of yelling people), but will want to know that bug is fixed before moving to trunk lettuce.

skoczen commented:
Alright, all comments in on the PR!   Would ideally like to see all of these and the ones by netme either fixed or addressed with why they shouldn't be fixed.

Apologies there are so many - that, I'm sure, feels like a huge burden.  This is one of the biggest reasons to keep PRs small and focused on a single feature.  It keeps these lists of changes at bay.

As for the bigger picture, here's what I've got, for this to be merged:

### Zope.testbrowser

If it's totally busted, ok.  But it installs fine over here.  It's useful because it's a browser with javascript disabled (as some users have their browsers set up), and is helpful for covering those use-cases.

### $ at the end of regexes.

This is a breaking change, and I'm not comfortable just dropping it on people.  Since many people don't have salad locked to a specific version, just running `pip install --upgrade` will break their tests.  

I know I have projects that actually *use* the non-strictness of the regexes as a feature for more readable tests.

Here's what I propose instead.

1. Add a `--strict` flag, that turns on the `$` at the end of *ALL* line-ending regexes. If the flag is on, the regexes include the `$`. If it's off, they continue as they are right now.
2.  The addition of `$`s is not (as far as I can see) on all relevant regexes. To add this, they must be everywhere or nowhere.
2. Document the flag, both in the CLI and the README.

### The new storage functions.

I'm pretty iffy on these - they feel really domain-specific, and open a can of worms (for instance, why not include [all of these](https://github.com/GoodCloud/django-functional-tests/blob/master/functional_tests/factory.py) sorts of generators?)  but given that they're useful to you all, I'm willing to let them go in.

However, they do need really clear documentation in the readme on how to use them, what they're for, when the stored values go away, etc.   A few examples showing them in a best-case practice would also be great.

### All new matchers need documentation

There are a lot of new matchers that need to be documented in the README.  Our goal here is that nobody picking up and using salad has to dig through the code to understand what they can do.  The README should be everything.

That's my (admittedly large) list!

Thanks again for all your efforts on this, and please post any questions/clarifications/thoughts you have - let's keep this an open dialog!
